### PR TITLE
feat(bin, storage): configurable MDBX log level

### PR DESCRIPTION
**Pull request by shekhirin**
As per the PR title.

I don't particularly like how we pass `log_level` as an argument to `init_db/open_db/etc` functions, but there seems to be no way around it until we get rid of helper functions and introduce a proper `Database::new` or something.

MDBX log levels higher than `Notice` require a debug build:
```console
Database:
      --db.log-level <LOG_LEVEL>
          Database logging level. Levels higher than "notice" require a debug build

          Possible values:
          - fatal:   Enables logging for critical conditions, i.e. assertion failures
          - error:   Enables logging for error conditions
          - warn:    Enables logging for warning conditions
          - notice:  Enables logging for normal but significant condition
          - verbose: Enables logging for verbose informational
          - debug:   Enables logging for debug-level messages
          - trace:   Enables logging for trace debug-level messages
          - extra:   Enables logging for extra debug-level messages

```

---

Debug build:
```console
➜  reth (alexey/mdbx-log-level) cargo run --quiet db --db.log-level trace stats
uniq_check:5744 uniq-probe: unique
setup_lck:18201 lck-setup: with-lck readonly exclusive
osal_mmap:31938 ftruncate 1032192, err 0
osal_mmap:32057 UNPOISON_MEMORY_REGION(0x10e654000, 1032192) at 32057
read_header:16396 reading meta[0]: offset 0, bytes 256, retry-left 42
validate_meta:16147 meta[0] took pagesize 16384
validate_meta:16172 checking meta0 = root 9214/9230, geo 3/10066-262144/268435456 +262144 -524288, txn_id 70, Steady
...
```

Release build:
```console
➜  reth (alexey/mdbx-log-level) cargo run --quiet --release db --db.log-level notice stats
setup_dxb:18079 open-MADV_DONTNEED 10064..262144
set_readahead:9839 readahead OFF 0..10064
```

Release build with unavailable log level:
```console
➜  reth (alexey/mdbx-log-level) cargo run --quiet --release db --db.log-level trace stats
Error: Could not open database at path: /Users/shekhirin/Library/Application Support/reth/mainnet/db

Caused by:
    Log level is not available: Trace

Location:
    crates/storage/db/src/lib.rs:140:14
```

### CODE REVIEW

**Review by rkrasiuk — COMMENTED** [review]
nice! however, i think that the CLI flag would become ambiguous for production builds (with `debug_assertions` disabled)
https://github.com/paradigmxyz/reth/blob/8233238f6afd9f500a568163e599febe50475077/crates/storage/libmdbx-rs/mdbx-sys/build.rs#L94-L98

**Comment by shekhirin** [comment]
> nice! however, i think that the CLI flag would become ambiguous for production builds (with `debug_assertions` disabled)
> 
> https://github.com/paradigmxyz/reth/blob/8233238f6afd9f500a568163e599febe50475077/crates/storage/libmdbx-rs/mdbx-sys/build.rs#L94-L98

yeah, let's also make enum variants which are not available in release build gated by `debug_assertions`, so we would also get less of them in `--help`

**Comment by gakonst** [comment]
We have `ProviderFactory::new_with_path`, should we pull in more stuff there and settle with the `ProviderFactory` as our entrypoint to all DB related work?

**Comment by gakonst** [comment]
Do we want to merge this, or is there more discussion per @rkrasiuk comment re: debug assertions?

**Review by rkrasiuk — APPROVED** [review]
fix seems good, ty

File: bin/reth/src/args/database_args.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+use reth_interfaces::db::LogLevel;
+
+/// Parameters for database configuration
+#[derive(Debug, Args, PartialEq, Default, Clone, Copy)]
+#[command(next_help_heading = "Database")]
+pub struct DatabaseArgs {
+    /// Database logging level. Levels higher than "notice" require a debug build.
+    #[arg(long = "db.log-level", value_enum)]
+    pub log_level: Option<LogLevel>,
+}
```
Comment (gakonst) [inline bin/reth/src/args/database_args.rs:13]: We'll probably want to move the DB Geometry config etc. here too in the future if it ends up being an issue?

File: crates/interfaces/src/db.rs [hunk 2]
```diff
@@ -31,4 +31,29 @@ pub enum DatabaseError {
...
     Stats(i32),
+    /// Failed to use the specified log level, as it's not available.
+    #[error("Log level is not available: {0:?}")]
+    LogLevelUnavailable(LogLevel),
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "clap", derive(clap::ValueEnum))]
+/// Database log level.
+pub enum LogLevel {
```
Comment (gakonst) [inline crates/interfaces/src/db.rs:42]: What shoudl be the default?

**Review by gakonst — APPROVED** [review]
supportive - feel free to merge

**Comment by shekhirin** [comment]
> We have `ProviderFactory::new_with_path`, should we pull in more stuff there and settle with the `ProviderFactory` as our entrypoint to all DB related work?

I think not every usage of `init_db` is straightforwardly replaceable with `ProviderFactory::new_with_database_path`, as we use a lot of `Database::view` and `Database::update` throughout the CLI tools. And it would be ideal IMO to upgrade MDBX log level even in CLI.

File: crates/interfaces/src/db.rs (hunk 2 above)
Comment (shekhirin) [reply crates/interfaces/src/db.rs:42]: `Fatal` is the default set by MDBX itself:
https://github.com/paradigmxyz/reth/blob/ce50db0b1d65f959b11d064157dd56696e282d30/crates/storage/libmdbx-rs/mdbx-sys/libmdbx/mdbx.c#L7247

I think it doesn't make sense for us to always call `mdbx_setup_debug` even if the log level doesn't change from `Fatal`.

**Pull request merged**